core/src/main/java/io/questdb/griffin/engine/functions/str/SplitPartCharFunctionFactory.java

```diff
@@ -0,0 +1,252 @@
+import org.jetbrains.annotations.Nullable;
+
+public class SplitPartCharFunctionFactory implements FunctionFactory {
+    @Override
+    public String getSignature() {
+        return "split_part(SAI)";
+    }
+
+    @Override
+    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) throws SqlException {
+        final Function strFunc = args.getQuick(0);
+        final Function delimiterFunc = args.getQuick(1);
+        final Function indexFunc = args.getQuick(2);
+
+        ArgFunctions argFunctions = new ArgFunctions();
+        ArgFunctions.Arg<CharSequence> strArg = argFunctions.registerDynamicArg(strFunc, "str", argPositions.getQuick(0), func -> func::getStr);
+        ArgFunctions.Arg<Character> delimiterArg = argFunctions.registerDynamicArg(delimiterFunc, "delimiter", argPositions.getQuick(1), func -> func::getChar);
+        ArgFunctions.Arg<Integer> indexArg = argFunctions.registerRuntimeConstantArg(indexFunc, "index", argPositions.getQuick(2), func -> func::getInt);
+
+        Function fallback = ArgFunctions.validateArg(delimiterArg, (delimiter, argName, argPosition) -> {
+            if (delimiter == CharConstant.ZERO.getChar(null)) {
+                return StrConstant.NULL;
+            }
+            return null;
+        }, null);
+        if (fallback != null) {
+            return fallback;
+        }
+
+        fallback = ArgFunctions.validateArg(indexArg,
+                (index, argName, argPosition) -> {
+                    if (index == Numbers.INT_NaN) {
+                        return StrConstant.NULL;
+                    } else if (index == 0) {
+                        throw SqlException.$(argPosition, "field position must not be zero");
+                    }
+                    return null;
+                }, (index, argName, argPosition) -> {
+                    if (index == 0) {
+                        throw SqlException.$(argPosition, "field position must not be zero");
+                    }
+                });
+        if (fallback != null) {
+            return fallback;
+        }
+
+        int argCount = argFunctions.getDynamicArgs().size();
+        if (argCount == 3) {
+            return new SplitPartTernaryFunction(argFunctions, strArg, delimiterArg, indexArg);
+        } else if (argCount == 2) {
+            return new SplitPartBinaryFunction(argFunctions, strArg, delimiterArg, indexArg);
+        } else if (argCount == 1) {
+            return new SplitPartUnaryFunction(argFunctions, strArg, delimiterArg, indexArg);
+        } else if (argCount == 0) {
+            return new SplitPartFunction(argFunctions, strArg, delimiterArg, indexArg);
+        } else {
+            throw new RuntimeException("Unexpected number of args: " + argCount);
+        }
+    }
+
+    private static class SplitPartTernaryFunction extends SplitPartFunction implements TernaryFunction {
```

Comment:
This class seems redundant, just like the next one: `indexArg.getValue(rec)` calls are made in any case. It's fine to keep `SplitPartFunction` class only.

Reply:
The difference is that they implement `UnaryFunction`/`BinaryFunction`/`TernaryFunction`. For example, if both `delimiter` and `index` args are constant or runtime constant, then a `SplitPartUnaryFunction` is returned. If none of them are constant or runtime constant, then a `SplitPartTernaryFunction` is returned. If either `delimiter` or `index` args are constant or runtime constant, then a `SplitPartBinaryFunction` is returned. This avoids having separate classes for constant `delimiter` vs constant `index` args, the same `SplitPartBinaryFunction` class can be reused.

Reply:
> The difference is that they implement `UnaryFunction`/`BinaryFunction`/`TernaryFunction`.

These helper interfaces aren't valuable on their own, i.e. it doesn't make a lot of difference whether a function implements `UnaryFunction` or `BinaryFunction` from the performance point of view.

Reply:
Maybe I misunderstood the purpose of these interfaces (perhaps they should be documented somewhere). Does that mean it would be fine to return `TernaryFunction` in all cases (regardless if the args are constant or not)?

Reply:
> Does that mean it would be fine to return `TernaryFunction` in all cases (regardless if the args are constant or not)?

Yes.

Reply:
Ok I've changed it to return `TernaryFunction` in all cases. This also simplifies the utility class as it no longer needs to track the number of dynamic arg functions.